### PR TITLE
feat: enable bypassing persisted operations

### PR DESCRIPTION
**Pull request by leo91000**
## Description

This PR is a feature originally discussed in #2 

It allows for an additional postgraphile option `allowUnpersistedQueries (req: any): boolean` for bypassing persisted queries (you usually want that in development environment).

## Performance impact

This feature adds code that is executed if the query hash is not found. It does the following :
- Get allowUnpersistedQueries
- If allowUnpersistedQueries returns true, it get the query from payload and returns it

## Security impact

Maybe we should add a warning for developpers to inform that an unpersisted query has been sent ?

## Checklist

- [x] My code matches the project's code style and `yarn lint:fix` passes.
- [ ] I've added tests for the new feature, and `yarn test` passes.
- [x] I have detailed the new feature in the relevant documentation.
- [x] I have added this feature to 'Pending' in the `RELEASE_NOTES.md` file (if one exists).
- [x] If this is a breaking change I've explained why.

### CODE REVIEW

File: README.md
Comment (benjie) [inline README.md]:
```suggestion
 *
 * @example
 *
 * ```
 * app.use(postgraphile(DATABASE_URL, SCHEMAS, {
 *   allowUnpersistedOperation(req) {
 *     return process.env.NODE_ENV === "development" && req.headers.referer?.endsWith("/graphiql");
 *   }
 * });
 * ```
```

File: RELEASE_NOTES.md
Comment (benjie) [inline RELEASE_NOTES.md]:
```suggestion
- Added `allowUnpersistedOperation` option, allowing arbitrary operations to be issued under controlled circumstances
```

File: README.md
Comment (benjie) [inline README.md]:
```suggestion
 * There are situations where you may want to allow arbitrary operations
 * (for example using GraphiQL in development, or allowing an admin to
 * make arbitrary requests in production) whilst enforcing Persisted
 * Operations for the application and non-admin users. This function
 * allows you to determine under which circumstances persisted operations
 * may be bypassed.
```

File: src/index.ts
Comment (benjie) [inline src/index.ts]: (As above)

File: src/index.ts
Comment (benjie) [inline src/index.ts]: The payload being passed here is not the `req` object; it should be. (We should also pass the payload as a second option.)

The line `"postgraphile:httpParamsList"(paramsList, { options }) {` below I think can get access to `req`: `"postgraphile:httpParamsList"(paramsList, { options, req }) {`

File: src/index.ts
Comment (benjie) [inline src/index.ts]: Let's be more direct; we're ultimately doing `params.query = params.query` after all.

```suggestion
        if (typeof payload.query === "string") {
          return payload.query;
        }
```

File: README.md
Comment (benjie) [inline README.md]: I think the type for this is something like `IncomingMessage`? You'll also want the payload here:

```suggestion
allowUnpersistedOperation?(request: IncomingMessage, payload: any}): boolean;
```

**Review by benjie — CHANGES_REQUESTED** [review]
Excellent start :raised_hands:

File: src/index.ts
Comment (leo91000) [reply src/index.ts]: I have adapted the code to use the real `req` object, is it ok ? 

Also, I'm not sure if it works with websocket too (`"postgraphile:ws:onOperation"(paramsList, { options, req }) {`)

File: RELEASE_NOTES.md
Comment (benjie) [inline RELEASE_NOTES.md]: Please move this above the `### v0.0.1` entry; thanks!

File: src/index.ts
Comment (benjie) [inline src/index.ts]: This puts unnecessary load on the garbage collector by creating a lambda function that'll be thrown away moments later. Better to solve it below:

File: src/index.ts
Comment (benjie) [inline src/index.ts]:
```suggestion
        options.allowUnpersistedOperations ? options.allowUnpersistedOperations(req, params) : false
```

File: src/index.ts
Comment (benjie) [inline src/index.ts]: We're not sent the `req` here; with hooks you have to search to see where they're called and what they're passed:

https://github.com/graphile/postgraphile/blob/ba9a7bcf1c3fa3347cf07de0a3732cfd0d5b6dcb/src/postgraphile/http/subscriptions.ts#L261-L266

However, we do have a helper for that:

https://github.com/graphile/postgraphile/blob/ba9a7bcf1c3fa3347cf07de0a3732cfd0d5b6dcb/src/postgraphile/http/subscriptions.ts#L405

which it turns out extracts req via:

```
    const req = socket['__postgraphileReq'];
```

So you should be able to get a req you can use:

```suggestion
  "postgraphile:ws:onOperation"(params, { message, options, socket }) {
    const req = socket['__postgraphileReq'];
```

**Review by benjie — DISMISSED** [review]
Looking good; almost there :+1:

File: src/index.ts
Comment (leo91000) [reply src/index.ts]: Thanks 👍🏼 I wouldn't have guessed without your help.

File: RELEASE_NOTES.md
Comment (leo91000) [reply RELEASE_NOTES.md]: Did you mean above `### v0.0.3` ?

File: src/index.ts
Comment (benjie) [inline src/index.ts]: It might make sense to allow `allowUnpersistedOperations: true` as the setting; so this check would become:

```suggestion
        typeof options.allowUnpersistedOperations === "function"
          ? options.allowUnpersistedOperations(req, params)
          : !!options.allowUnpersistedOperations
```

File: src/index.ts
Comment (benjie) [inline src/index.ts]: As below, we could have this accept a boolean too:

```suggestion
    allowUnpersistedOperation?: boolean | ((request: IncomingMessage, payload: any) => boolean);
```

(This change would need copying to the README also.)

**Review by benjie — COMMENTED** [review]
This looks great. I'll need to test it before I can merge it. What do you think about enabling it to receive a boolean as well as a function?

**Comment by leo91000** [comment]
@benjie I added the option to pass directly a boolean

File: README.md
Comment (benjie) [inline README.md:71]: This was likely Prettier escaping a code block inside a code block; I don't think you need it so please go back to just 3 backticks
````suggestion
```ts
````

File: README.md
Comment (benjie) [inline README.md:133]:
````suggestion
```
````

File: README.md
Comment (benjie) [inline README.md:118]:
```suggestion
 * may be bypassed.
 *
 * IMPORTANT: this function must not throw!
```

File: src/index.ts
Comment (benjie) [inline src/index.ts:61]:
```suggestion
     * may be bypassed.
     *
     * IMPORTANT: this function must not throw!
```

File: src/index.ts
Comment (benjie) [inline src/index.ts]: Slightly simplified, and won't throw error if `options.allowUnpersistedOperation` is a string, object or array:

```suggestion
  const { allowUnpersistedOperation } = options;
  if (typeof allowUnpersistedOperation === 'function') {
    return allowUnpersistedOperation(request, payload);
  }
  return !!allowUnpersistedOperation;
```

**Review by benjie — CHANGES_REQUESTED** [review]
Great; I think I might want to revise the type of `payload` from `any` to something a bit tighter; but other than that this looks solid :+1: 

Leave it with me now.

**Review by benjie — APPROVED** [review]

**Comment by leo91000** [comment]
Thanks, you are awesome ! 💯

**Comment by benjie** [comment]
(I haven't forgotten about this.)

**Comment by leo91000** [comment]
> (I haven't forgotten about this.)

No problem take your time

**Comment by benjie** [comment]
Hey @leo91000 can you grant me permission to push to this branch?

**Comment by leo91000** [comment]
> Hey @leo91000 can you grant me permission to push to this branch?

I am not sure how to achieve that 🤔. I invited you to the repository, is this the way ?

**Comment by benjie** [comment]
That works; in future you can just check a checkbox in the right sidebar "Allow edits from maintainers":

https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Review by benjie — APPROVED** [review]

**Comment by benjie** [comment]
Release in v0.1.0 :tada:

**Comment by leo91000** [comment]
Yay 🎉 ! Thank you for taking time for this !

**Comment by benjie** [comment]
Thanks for writing it; it was the remaining thing stopping me from moving it to 0.x from 0.0.x :grin:

**Pull request merged**
